File: finance-ui/src/app/services/wallet.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { AuthService } from './auth.service';

export interface WalletResponse {
  id: number;
  userId: number;
  balance: number;
  currency: 'TRY' | 'USD' | 'EUR';
  status: 'ACTIVE' | 'PENDING' | 'BLOCKED';
  createdDate: string;
  updatedDate: string;
}

@Injectable({
  providedIn: 'root'
})
export class WalletService {
  private readonly http = inject(HttpClient);
  private readonly authService = inject(AuthService);
  private readonly baseUrl = '/wallets/api/v1';

  private getAuthHeaders(): { headers: HttpHeaders } {
    const token = this.authService.getToken();
    return {
      headers: new HttpHeaders({
        'Authorization': `Bearer ${token}`
      })
    };
  }

  getWalletByUserId(userId: number): Observable<WalletResponse> {
    return this.http.get<WalletResponse>(`${this.baseUrl}/user/${userId}`, this.getAuthHeaders());
  }

  createWallet(userId: number, initialBalance: number = 0, currency: string = 'TRY', status: string = 'ACTIVE'): Observable<WalletResponse> {
    const body = {
      userId,
      balance: initialBalance,
      currency,
      status
    };
    return this.http.post<WalletResponse>(`${this.baseUrl}/create`, body, this.getAuthHeaders());
  }

  deposit(walletId: number, amount: number): Observable<WalletResponse> {
    return this.http.post<WalletResponse>(`${this.baseUrl}/${walletId}/deposit?amount=${amount}`, {}, this.getAuthHeaders());
  }

  withdraw(walletId: number, amount: number): Observable<WalletResponse> {
    return this.http.post<WalletResponse>(`${this.baseUrl}/${walletId}/withdraw?amount=${amount}`, {}, this.getAuthHeaders());
  }

  transfer(sourceWalletId: number, destinationWalletId: number, amount: number): Observable<void> {
    return this.http.post<void>(
      `${this.baseUrl}/transfer?sourceWalletId=${sourceWalletId}&destinationWalletId=${destinationWalletId}&amount=${amount}`,
      {},
      this.getAuthHeaders()
    );
  }

  getTransactionHistory(walletId: number, page: number = 0, size: number = 20): Observable<any> {
    return this.http.get<any>(`${this.baseUrl}/${walletId}/history?page=${page}&size=${size}`, this.getAuthHeaders());
  }
}

File: my_finance_mobile_ui/lib/services/wallet_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/models.dart';
import 'api_config.dart';

class WalletService {
  final String _walletUrl = '${ApiConfig.baseUrl}/wallets/api/v1';

  // Helper to generate Authorization headers
  Map<String, String> _getHeaders(String token) {
    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $token',
    };
  }

  // Get Wallet by User ID
  Future<Wallet?> getWalletByUserId(int userId, String token) async {
    final response = await http.get(
      Uri.parse('$_walletUrl/user/$userId'),
      headers: _getHeaders(token),
    );

    if (response.statusCode == 200) {
      if (response.body.isEmpty) return null;
      final decodedJson = jsonDecode(utf8.decode(response.bodyBytes));
      return Wallet.fromJson(decodedJson);
    } else if (response.statusCode == 404) {
      return null;
    } else {
      throw Exception('Cüzdan bilgileri alınamadı.');
    }
  }

  // Create standard Wallet
  Future<Wallet> createWallet({
    required int userId,
    required double initialBalance,
    required Currency currency,
    required String token,
  }) async {
    final response = await http.post(
      Uri.parse('$_walletUrl/create'),
      headers: _getHeaders(token),
      body: jsonEncode({
        'userId': userId,
        'balance': initialBalance,
        'currency': currency.name,
        'status': 'ACTIVE',
      }),
    );

    if (response.statusCode == 200 || response.statusCode == 201) {
      final decodedJson = jsonDecode(utf8.decode(response.bodyBytes));
      return Wallet.fromJson(decodedJson);
    } else {
      String errorMessage = 'Cüzdan oluşturulamadı.';
      try {
        final errorBody = jsonDecode(utf8.decode(response.bodyBytes));
        errorMessage = errorBody['message'] ?? errorMessage;
      } catch (_) {}
      throw Exception(errorMessage);
    }
  }

  // Deposit money into wallet
  Future<Wallet> deposit({
    required int walletId,
    required double amount,
    required String token,
  }) async {
    final response = await http.post(
      Uri.parse('$_walletUrl/$walletId/deposit?amount=$amount'),
      headers: _getHeaders(token),
    );

    if (response.statusCode == 200) {
      final decodedJson = jsonDecode(utf8.decode(response.bodyBytes));
      return Wallet.fromJson(decodedJson);
    } else {
      String errorMessage = 'Para yatırılamadı.';
      try {
        final errorBody = jsonDecode(utf8.decode(response.bodyBytes));
        errorMessage = errorBody['message'] ?? errorMessage;
      } catch (_) {}
      throw Exception(errorMessage);
    }
  }

  // Withdraw money from wallet
  Future<Wallet> withdraw({
    required int walletId,
    required double amount,
    required String token,
  }) async {
    final response = await http.post(
      Uri.parse('$_walletUrl/$walletId/withdraw?amount=$amount'),
      headers: _getHeaders(token),
    );

    if (response.statusCode == 200) {
      final decodedJson = jsonDecode(utf8.decode(response.bodyBytes));
      return Wallet.fromJson(decodedJson);
    } else {
      String errorMessage = 'Para çekilemedi.';
      try {
        final errorBody = jsonDecode(utf8.decode(response.bodyBytes));
        errorMessage = errorBody['message'] ?? errorMessage;
      } catch (_) {}
      throw Exception(errorMessage);
    }
  }

  // Transfer money to another wallet ID
  Future<void> transfer({
    required int sourceWalletId,
    required int destinationWalletId,
    required double amount,
    required String token,
  }) async {
    final response = await http.post(
      Uri.parse(
          '$_walletUrl/transfer?sourceWalletId=$sourceWalletId&destinationWalletId=$destinationWalletId&amount=$amount'),
      headers: _getHeaders(token),
    );

    if (response.statusCode != 200 && response.statusCode != 204) {
      String errorMessage = 'Transfer gerçekleştirilemedi.';
      try {
        final errorBody = jsonDecode(utf8.decode(response.bodyBytes));
        errorMessage = errorBody['message'] ?? errorMessage;
      } catch (_) {}
      throw Exception(errorMessage);
    }
  }

  // Get transaction history (returns List of Transactions from Spring Page structure)
  Future<List<Transaction>> getTransactionHistory({
    required int walletId,
    required String token,
    int page = 0,
    int size = 20,
  }) async {
    final response = await http.get(
      Uri.parse('$_walletUrl/$walletId/history?page=$page&size=$size'),
      headers: _getHeaders(token),
    );

    if (response.statusCode == 200) {
      final decodedJson = jsonDecode(utf8.decode(response.bodyBytes));
      final List<dynamic> content = decodedJson['content'] as List<dynamic>? ?? [];
      return content.map((tx) => Transaction.fromJson(tx)).toList();
    } else {
      throw Exception('İşlem geçmişi yüklenemedi.');
    }
  }
}
